src/sdl/index.ts

```diff
@@ -142,6 +137,127 @@ export class SDL {
     }
   }
 
+  private readonly ENDPOINT_NAME_VALIDATION_REGEX = /^[a-z]+[-_\da-z]+$/;
+
+  private readonly ENDPOINT_KIND_IP = "ip";
+
+  private readonly endpointsUsed = new Set<string>();
+
+  private readonly portsUsed = new Map<string, string>();
+
+  constructor(
+    public readonly data: v2Sdl,
+    public readonly version: NetworkVersion = "beta2",
+    private readonly networkId: NetworkId = MAINNET_ID
+  ) {
+    this.validate();
+  }
+
+  private validate() {
+    // TODO: this should really be cast to unknown, then assigned
+    // to v2 or v3 SDL only after being validated
+    const v3data = this.data as v3Sdl;
+    Object.entries(v3data.profiles.compute).forEach(([name, { resources }]) => {
+      if ("gpu" in resources) {
+        SDL.validateGPU(name, resources.gpu);
+      }
+      SDL.validateStorage(name, resources.storage);
+    });
+
+    this.validateEndpoints();
+
+    Object.keys(this.data.services).forEach(serviceName => {
+      this.validateDeploymentWithRelations(serviceName);
+      this.validateLeaseIP(serviceName);
+      this.validateCredentials(serviceName);
+    });
+
+    this.validateDenom();
+  }
+
+  private validateDenom() {
+    const usdcDenom = USDC_IBC_DENOMS[this.networkId];
+    const denoms = this.groups()
+      .flatMap(g => g.resources)
+      .map(resource => resource.price.denom);
+    const invalidDenom = denoms.find(denom => denom !== "uakt" && denom !== usdcDenom);
+
+    CustomValidationError.assert(!invalidDenom, `Invalid denom: "${invalidDenom}". Only uakt and ${usdcDenom} are supported.`);
+  }
+
+  private validateEndpoints() {
+    if (!this.data.endpoints) {
+      return;
+    }
+
+    Object.keys(this.data.endpoints).forEach(endpointName => {
+      const endpoint = this.data.endpoints[endpointName];
+      CustomValidationError.assert(this.ENDPOINT_NAME_VALIDATION_REGEX.test(endpointName), `Endpoint named "${endpointName}" is not a valid name.`);
+      CustomValidationError.assert(!!endpoint.kind, `Endpoint named "${endpointName}" has no kind.`);
+      CustomValidationError.assert(endpoint.kind === this.ENDPOINT_KIND_IP, `Endpoint named "${endpointName}" has an unknown kind "${endpoint.kind}".`);
+    });
+  }
+
+  private validateCredentials(serviceName: string) {
+    const { credentials } = this.data.services[serviceName];
+
+    if (credentials) {
+      const credentialsKeys: (keyof v2ServiceImageCredentials)[] = ["host", "username", "password"];
```

Comment:
I think `email` is missing https://github.com/akash-network/akash-api/blob/e82f3d1884a7760a90b5e8784db9ccf25cbdbb65/proto/provider/akash/manifest/v2beta2/service.proto#L39

Reply:
It is there in type, however its not validated in node https://github.com/akash-network/node/commit/6e68af006959da478a4cf1af1ef0b663347852c5#diff-346f62ceedbdbca807c73a4314af6b8da932fdb8ee2e91c92dbecd5c2745f24aR190

so i skipped any validation here too

Reply:
so is this fine @baktun14 ?

Reply:
Ah ok I guess yea, this is intended? @troian